Remotion/Data/DomainObjects/Mapping/Builder/ClassDefinitionBuilderNode.cs

```diff
@@ -25,26 +27,48 @@ namespace Remotion.Data.DomainObjects.Mapping.Builder
   /// Represents <see cref="Mapping.ClassDefinition"/> node in a tree constructed by <see cref="TypeDefinitionFactory"/>.
   /// Supports child navigation and correctly setting up child relationships in a resulting <see cref="Mapping.ClassDefinition"/>.
   /// </summary>
+  [DebuggerDisplay("Builder node for class '{Type}'")]
   public class ClassDefinitionBuilderNode : TypeDefinitionBuilderNode
   {
     public ClassDefinitionBuilderNode? BaseClass { get; }
 
+    public IReadOnlyList<InterfaceDefinitionBuilderNode> ImplementedInterfaces { get; }
+
+    public bool IsInheritanceRoot { get; }
+
     public ClassDefinition? ClassDefinition { get; private set; }
 
     public bool IsConstructed { get; private set; }
 
     private readonly List<ClassDefinitionBuilderNode> _derivedClasses = new();
 
-    public ClassDefinitionBuilderNode (Type type, ClassDefinitionBuilderNode? baseClass)
+    public ClassDefinitionBuilderNode (Type type, ClassDefinitionBuilderNode? baseClass, IEnumerable<InterfaceDefinitionBuilderNode> implementedInterfaces)
         : base(type)
     {
+      ArgumentUtility.CheckNotNull("implementedInterfaces", implementedInterfaces);
       if (!type.IsClass)
         throw new ArgumentException("The specified type must be a class.", "type");
+      if (type == typeof(DomainObject))
+        throw new ArgumentException("Cannot create a builder node for the DomainObject type.", "type");
+      if (!ReflectionUtility.IsDomainObject(type))
+        throw new ArgumentException("Cannot create a builder node for a type that is not a domain object.", "type");
       if (baseClass?.IsConstructed == true)
         throw new ArgumentException("The specified base class must not be a constructed node.", "baseClass");
 
+      var implementedInterfacesArray = implementedInterfaces.ToArray();
+      if (implementedInterfacesArray.Any(e => e.IsConstructed))
+        throw new ArgumentException("The specified implemented interfaces must not be constructed.", "implementedInterfaces");
+
+      IsInheritanceRoot = ReflectionUtility.IsInheritanceRoot(type);
+
       BaseClass = baseClass;
+      ImplementedInterfaces = IsInheritanceRoot || baseClass == null
```

Comment:
```suggestion
      ImplementedInterfaces = baseClass == null
```
This should be identical?

Unless this is because baseClass can become not-null during processing. If that makes a difference for evaluating ImplementedInterfaces, I think we have a design problem because the code becomes strange.

Reply:
No we need the check for `IsInheritanceRoot` here. Initially the node tree is built using just the base type/interface types information. So for an inheritance root it is possible that `baseClass != null`. In such a case we want to pull all the interfaces onto the inheritance root as the base class is not linked afterwards.